subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java

```diff
@@ -145,6 +149,20 @@ public FileCollection getOutputFiles() {
         }
 
         @Override
+        public Set<File> getPreviousOutputs() {
+            TaskExecution previousExecution = history.getPreviousExecution();
+            if (previousExecution == null || previousExecution.getOutputFilesSnapshot() == null) {
+                return Collections.emptySet();
+            }
+            ImmutableCollection<FileCollectionSnapshot> outputFilesSnapshot = previousExecution.getOutputFilesSnapshot().values();
+            Set<File> outputs = new HashSet<File>();
+            for (FileCollectionSnapshot fileCollectionSnapshot : outputFilesSnapshot) {
+                outputs.addAll(fileCollectionSnapshot.getElements());
+            }
+            return outputs;
+        }
```

Comment:
We already talk about `StaleClassCleaner` - this is still required since it is used to clean outputs of the same task when the task is not incremental.
There is now only one implementation of the method returning a `Set<File>` and directories and files.